src/hnsw.c

```diff
@@ -119,15 +121,53 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
 	HnswGetMetaPageInfo(index, &m, NULL);
 	index_close(index, NoLock);
 
-	/* Approximate entry level */
-	entryLevel = (int) -log(1.0 / path->indexinfo->tuples) * HnswGetMl(m);
+	/*
+	 * HNSW cost estimation follows a formula that accounts for the total
+	 * number of tuples indexed combined with the parameters that most influence
+	 * the duration of the index scan, namely:
+	 *   m - the number of tuples that are scanned in each step of the HNSW
+	 *       graph traversal
+	 *   ef_search - which influences the total number of steps taken at layer 0
+	 *
+	 *  The source of the vector data can impact how many steps it takes to
+	 *  converge on the set of vectors to return to the executor. Currently,
+	 *  we use a hardcoded scaling factor (HNSWScanScalingFactor) to help
+	 *  influence that, but this could later become a configurable parameter
+	 *  based on the cost estimations.
+	 *
+	 * The tuple estimator formula is below:
+	 *
+	 * numIndexTuples = (totalLayers * m) +
+	 * 					(layer0TuplesMax * layer0Selectivity * HnswSF)
+	 *
+	 * "totalLayers * m" represents the floor of tuples we need to scan to get
+	 * to layer 0 (L0).
+	 *
+	 * "layer0TuplesMax" is the estimated total number of tuples we'd scan at
+	 * L0 if we weren't discarding already visited tuples as part of the scan.
+	 *
+	 * "layer0Selectivity" estimates the percentage of tuples that are scanned
+	 * at L0, accounting for previously visited tuples.
+	 *
+	 * "HnswSF" is a scaling factor, hardcoded to HnswScanScalingFactor
+	 */
+	totalLayers = floor(log(path->indexinfo->tuples) / log(m));
```

Comment:
I think it'd be better to keep the previous naming (`entryLevel`) here, as well as `HnswGetMl(m)` since it's really the total number of layers - 1.

We may also want to add `+ 1` to the number of tuples here and below to avoid `log(0)`.

Reply:
@ankane I'm less opinionated on the name; I did try to force `tuples` to 0, but in that case PostgreSQL won't consider the scan. I think in the case where `tuples` is 0, we should max out the costs like we do if there's no ORDER BY quals.